apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js

```diff
@@ -169,3 +174,120 @@ GoogleBlockly.Extensions.registerMutator(
+ * https://github.com/google/blockly/blob/5a23c84e6ef9c0b2bbd503ad9f58fa86db1232a8/core/procedures.ts#L202-L287
+ * @param {WorkspaceSvg} workspace The workspace containing procedures.
+ * @returns an array of block objects representing the flyout blocks
+ */
+export function flyoutCategory(workspace, functionEditorOpen = false) {
+  const useNewFunctionEditor = experiments.isEnabled(
+    experiments.MODAL_FUNCTION_EDITOR
+  );
+  const blockList = [];
+
+  const newFunctionButton = {
+    kind: 'button',
+    text: msg.createBlocklyFunction(),
+    // TODO: Remove the alternate callback key once we're using the new function editor
+    callbackKey: useNewFunctionEditor
+      ? 'newProcedureCallback'
+      : 'createNewFunction',
+  };
+  const functionDefinitionBlock = {
+    kind: 'block',
+    type: 'procedures_defnoreturn',
+    fields: {
+      NAME: Blockly.Msg.PROCEDURES_DEFNORETURN_PROCEDURE,
+    },
+  };
+
+  const createNewFunction = function () {
+    // Everything here is place-holder code that should be replaced with a
+    // call to open the behavior editor with a new defintion block.
+    // Until then, we just create a block under all existing blocks on the
+    // main workspace.
+    createNewDefinitionBlock(functionDefinitionBlock);
+  };
+
+  if (functionEditorOpen) {
+    // No-op - cannot create new functions while the modal editor is open
+  } else if (useModalFunctionEditor) {
+    workspace.registerButtonCallback('createNewFunction', createNewFunction);
+    workspace.registerButtonCallback(
+      'newProcedureCallback',
+      Blockly.functionEditor.newProcedureCallback
+    );
+    blockList.push(newFunctionButton);
+  } else {
+    blockList.push(functionDefinitionBlock);
+  }
+
+  // Workspaces to populate functions flyout category from
+  const workspaces = [
+    Blockly.getMainWorkspace(),
+    Blockly.getHiddenDefinitionWorkspace(),
+  ];
+
+  const allFunctions = [];
+  workspaces.forEach(workspace => {
+    const procedureBlocks = workspace
+      .getTopBlocks()
+      .filter(topBlock => topBlock.type === 'procedures_defnoreturn');
+    procedureBlocks.forEach(block => {
+      allFunctions.push({
+        name: block.getFieldValue('NAME'),
+        id: block.id,
+      });
+    });
+  });
+
+  allFunctions.sort(nameComparator).forEach(({name, id}) => {
+    blockList.push({
+      kind: 'block',
+      type: 'procedures_callnoreturn',
+      extraState: {
+        name: name,
+        id: id,
+      },
+      fields: {
+        NAME: name,
+      },
+    });
+  });
+
+  return blockList;
+}
+
+const getLowestBlockBottomY = () => {
+  let lowestBlockBottom = 0;
+  Blockly.getMainWorkspace()
+    .getTopBlocks()
+    .forEach(block => {
+      const blockY = block.getRelativeToSurfaceXY().y;
+      const blockBottom = blockY + block.getHeightWidth().height;
+      if (blockBottom > lowestBlockBottom) {
+        lowestBlockBottom = blockBottom;
+      }
+    });
+  return lowestBlockBottom + 16;
```

Comment:
why do we add 16 here? Can you add a comment explaining this?

Reply:
My understanding is this is an arbitrary offset for when we add a new definition block to the bottom of of the blockspace. I added a constant BLOCK_OFFSET at the top of the file since we use 16 both for x and y in `createNewDefinitionBlock`, so it's now:
  const newDefinitionBlock = Blockly.serialization.blocks.append(
    {...blockState, x: BLOCK_OFFSET, y: getLowestBlockBottomY() + BLOCK_OFFSET},
    Blockly.getMainWorkspace()
  );
It's worth noting that eventually, this'll get removed.